ocs2_ros_interfaces/src/command/TargetTrajectoriesJoystickPublisher.cpp

```diff
@@ -0,0 +1,117 @@
+/******************************************************************************************************/
+TargetTrajectoriesJoystickPublisher::TargetTrajectoriesJoystickPublisher(::rclcpp::Node::SharedPtr& nodeHandle, const std::string& topicPrefix,
+                                                                         const scalar_array_t& targetCommandLimits,
+                                                                         CommandLineToTargetTrajectories commandLineToTargetTrajectoriesFun)
+    : targetCommandLimits_(Eigen::Map<const vector_t>(targetCommandLimits.data(), targetCommandLimits.size())),
+      commandLineToTargetTrajectoriesFun_(std::move(commandLineToTargetTrajectoriesFun)),
+      node_(nodeHandle)  {
+  observationReceived_ = false;
+  // observation subscriber
+  auto observationCallback = [this](const ocs2_msgs::msg::MPCObservation::SharedPtr msg) {
+    std::lock_guard<std::mutex> lock(latestObservationMutex_);
+    latestObservation_ = ros_msg_conversions::readObservationMsg(*msg);
+    observationReceived_ = true;
+  };
+  observationSubscriber_ = nodeHandle->create_subscription<ocs2_msgs::msg::MPCObservation>(topicPrefix + "_mpc_observation", 1, observationCallback);
+
+  // Trajectories publisher
+  targetTrajectoriesPublisherPtr_.reset(new TargetTrajectoriesRosPublisher(nodeHandle, topicPrefix));
+
+  command_ << 0,0,0,0;
+  //subsribe to joystick
+  joysubscription_ = node_->create_subscription<sensor_msgs::msg::Joy>(
+    "joy", 10, std::bind(&TargetTrajectoriesJoystickPublisher::joyCallback, this, std::placeholders::_1));
+}
+
+/******************************************************************************************************/
+/******************************************************************************************************/
+/******************************************************************************************************/
+void TargetTrajectoriesJoystickPublisher::publishJoystickCommand(const std::string& commadMsg) {
+  while (rclcpp::ok()) {
+    Eigen::Vector4d joy_command = getLatestJoyCommand().cwiseMin(targetCommandLimits_).cwiseMax(-targetCommandLimits_);
+    
+    // get the latest observation
+    ::rclcpp::spin_some(node_);
+
+    if (observationReceived_) {
+      SystemObservation observation;
+      {
+        std::lock_guard<std::mutex> lock(latestObservationMutex_);
+        observation = latestObservation_;
+      }
+
+      // Rotate joystick command to world frame since that's what ocs2 expects
+      Eigen::Quaterniond curr_rot = Eigen::AngleAxisd(observation.state[9],Eigen::Vector3d::UnitZ())
+                                * Eigen::AngleAxisd(observation.state[10],Eigen::Vector3d::UnitY())
+                                * Eigen::AngleAxisd(observation.state[11],Eigen::Vector3d::UnitX());
+      joy_command.segment<3>(0) = curr_rot.matrix()*joy_command.segment<3>(0);
```

Comment:
Do we necessarily want to apply the full rotation? I think only rotating about the yaw axis would be sufficient.

Otherwise, if your body is pitched up/down for example (tracking error, holding a pose, or something else) then the desired walking path would ascend/descend off the ground.

Or does OCS2 handle all that for us internally?

Reply:
that's fair, the command is relative to current pose so I'll just drop the Z component!
I'm not sure we want to add new buttons to also modify the z-axis!